pkg/restore/restore.go

```diff
@@ -384,6 +388,16 @@ func (ctx *restoreContext) execute() (Result, Result) {
 		return warnings, errs
 	}
 
+	// TODO: Remove outer feature flag check to make this feature a default in Velero.
+	if features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag) {
+		if ctx.backup.Status.FormatVersion >= "1.1.0" {
+			if err := ctx.chooseAPIVersionsToRestore(); err != nil {
```

Comment:
Since this call goes down to [an APIGroup discovery call](https://github.com/vmware-tanzu/velero/pull/3133/files#diff-597d46a83a1ad70945ab451fa42ec1eb96dbcbf427949e93b3a87792883d11f3R165), we may want to re-run the `chooseAPIVersionsToRestore` function [after restoring CRDs](https://github.com/vmware-tanzu/velero/pull/3133/files#diff-e0498755be37b683528e2569667d67818bb676827c420aab49adc62cb6d47fe3R492)

This is because at this line, we'll only have groups in the API server that existed before restore. As CRDs are restored, they'll be added into the API server for discovery, thus the list of API Groups discover here will fall out of date.

As an example, if I'm restoring the CRD `beatles.rockband.io`, it will not be listed in the API Groups at this line. Instead, it wouldn't be added until L492 of this file, since it was added during the restore process, and the API server would have it before then.

This is a problem Velero ran into before adding the "lazy restore" at L492, in that custom resources would try to restore, but their CRD did not yet exist in the API server yet.

Please let me know if this is unclear - it's a tricky race condition and may be better diagrammed.

Reply:
Since this is a tricky to wrap my head around. Hopefully you or @brito-rafa can correct me if I'm mistaken.

As I understand it, EnableAPIGroupVersions does not restore CRDs. It only restores CRs. We assume the CRDs will be installed manually by the user before the restore OR the CRDs will be the newer versions that come with Kubernetes. The reason I think that is because the new CRDs won't be in the backups, only the old ones will be.

But given my assumption, it brings up a question: when are the old CRDs restored? Even if EnableAPIGroupVersions doesn't play a part in that process, it needs to be able to discover these restored CRDs during the process of choosing a version to restore.

Reply:
> when are the old CRDs restored?

Great question! That information isn't immediately obvious in just this file.

[By default, Velero will restore them as the first items out of the tarball](https://github.com/vmware-tanzu/velero/blob/main/pkg/cmd/server/server.go#L448).

Inside the main restore loop, [we check if a CRD was restore, and if so, update the discovery client's list of API Groups.](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L477-L482). For each CRD we restore, [we wait to see if the Kubernetes API server marks it as available](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1185-L1186).

So the full list of CRDs won't be restored to the API server until after that loop finishes with the `customresourcedefinition` set and moves on to the rest in the `restoreResourcePriorities` list.

Regarding new versions of CRDs, I think you're correct - they'll need to be present prior to the restore happening, since Velero doesn't have them. For builtin kinds/API groups, this isn't a big deal, as they'll already be there. For applications, it may be trickier as the user will need to make sure only the CRDs are defined prior to the restore.

Reply:
So, as part of our testing using [rockbands.music.example.io](https://github.com/brito-rafa/k8s-webhooks/tree/master/examples-for-projectvelero), the target cluster has already a newer version of CRD at that time of the velero restore. This is a realistic scenario because the cluster-admin would have onboarded a newer controller in charge of such CRD. In case the target cluster does not have the CRD installed, velero restore "business as usual" procedure will kick in and create them because they are part of the velero backup tarball.

See below the `velero restore describe` from a restore using this PR velero version on a cluster that has the newer CRD already in place. Note that velero attempts to restore the CRD and then skips realizing that the CRD already exists on the target cluster and it is different. The CRs are restored.  Does it make sense?

```bash
velero restore describe case-b-20210202-20210203150211
(...)
Warnings:
  Velero:     <none>
  Cluster:  could not restore, customresourcedefinitions.apiextensions.k8s.io "rockbands.music.example.io" already exists. Warning: the in-cluster version is different than the backed-up version.
  Namespaces: <none>
```

Reply:
So far, this PR has not touched anything on CRD backup/restore logic. For your point Nolan, I think makes sense to re-run the discovery after velero restore the CRDs.

Reply:
@brito-rafa You're right that it doesn't touch CRDs specifically, I just wanted to call out how the discovery is working and how that may interact with this change.

But yeah, that makes sense - Velero can't know about the new versions, but in terms of discovering old versions, we need to check after we restore the CRDs and/or look in the backup (which we're already doing).

Reply:
Very informative conversation. Thank you @nrb to the links to the code, it helped me to understand at the code level what was happening. And thanks @brito-rafa for sharing that warning. I hadn't made the connection before that is why the "could not restore" message was coming up.

Given that discovery is already called after CRDs are restored, and the EnableAPIGroups feature only affects the restore process for non-CRDs by changing the file path to the resource in the backup, I think this comment thread can be marked "Resolved". Feel free @nrb and @brito-rafa to unresolve if I missed something.